src/main/scala/async/Listener.scala

```diff
@@ -1,73 +1,178 @@
+  def acceptingListener[T](consumer: T => Unit) =
+    new Listener[T]:
+      val lock = null
+      def complete(data: T, source: Source[T]) = consumer(data)
+
+  /** Returns a simple [[Listener]] that always accepts the item and sends it to the consumer. */
+  inline def apply[T](consumer: T => Unit): Listener[T] = acceptingListener(consumer)
+
+  /** A special class of listener that forwards the inner listener through the given source.
+    * For purposes of [[Async.Source.dropListener]] these listeners are compared for equality
+    * by the hash of the source and the inner listener.
+    */
+  abstract case class ForwardingListener[T](src: Async.Source[?], inner: Listener[T]) extends Listener[T]
+
+  object ForwardingListener:
+    /** Create an empty [[ForwardingListener]] for equality comparison. */
+    def empty[T](src: Async.Source[?], inner: Listener[T]) = new ForwardingListener(src, inner):
+      val lock = null
+      override def complete(data: T, source: Async.Source[T]) = ???
+
+  /** The result of locking a single listener lock. */
   sealed trait LockResult
 
-  object Locked extends LockResult
+  /** We have completed locking the listener. It can now be `complete`d. */
+  case object Locked extends LockResult
 
-  object Failed extends LockResult
+  /** The listener is no longer available.
+    * It should be removed from the source, and any acquired locks by the source must be
+    * manually `release`d by the source itself.
+    */
+  case object Gone extends LockResult
 
-  trait SemiLock extends LockResult:
+  /** Locking is successful; however, there are more locks to be acquired. */
+  trait PartialLock extends LockResult:
+    /** The number of the next lock. */
     val nextNumber: Long
+    /** Attempt to lock the next lock. */
     def lockNext(): LockResult
 
-  type ReleaseBoundary = SemiLock | Locked.type
+  /** Points to a position on the lock chain, whose lock up until this point has been acquired,
+    * but no further.
+    */
+  type LockMarker = PartialLock | Locked.type
 
-  trait TopLock:
+  /** A lock required by a listener to be acquired before accepting values.
+    * Should there be multiple listeners that needs to be locked at the same time,
+    * they should be locked by larger-number-first.
+
+    * Some implementations are provided for ease of implementations:
+    * - For custom listener implementations involving locks: [[NumberedLock]] provides uniquely
+    * numbered locks.
+    * - For source transformation implementations: [[ListenerLockWrapper]] provides a ListenerLock instance
+    *   that only forwards the requests to the underlying lock. [[withLock]] is a convenient
+    *   `.map` for `[[ListenerLock]] | Null`.
+    */
+  trait ListenerLock:
+    /** The assigned number of the lock.
+      * If the listener holds inner listeners underneath that utilizes locks,
+      * it is **required** that [[selfNumber]] must be larger than any [[PartialLock.nextNumber]] of
```

Comment:
We have that special case of Listeners that employ some form of rejection without fitting in the NumberedListener model. They can copy the downstream number (or use some -1 if they don't have any downstream (locking) listener.
```suggestion
      * it is **required** that [[selfNumber]] must be greater or equal any [[PartialLock.nextNumber]] of
```